File: tries/lb-06-all-shortest-unique-prefixes.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Node {
 public:
  int freq;
  bool terminal;
  Node *chars[26];
  Node() {
    freq = 0;
    for (int i = 0; i < 26; i++) {
      chars[i] = NULL;
    }
    terminal = false;
  }
};
class Trie {
  Node *root;

 public:
  Trie() { root = new Node(); }
  int get_index(char c) { return c - 'a'; }
  void insert(string word) {
    int index;
    Node *temp = root;
    for (int i = 0; i < word.length(); i++) {
      char w = word[i];
      index = get_index(w);
      if (!temp->chars[index]) {
        temp->chars[index] = new Node();
      }
      temp = temp->chars[index];
      temp->freq += 1;
    }
    temp->terminal = true;
  }
  vector<string> unique_prefixes(string words[], int n) {
    vector<string> ans;
    int index;
    Node *temp;
    for (int i = 0; i < n; i++) {
      temp = this->root;
      for (int j = 0, k = 1; j < words[i].length(); j++, k++) {
        char w = words[i][j];
        index = get_index(w);
        if (temp->chars[index]) {
          if (temp->chars[index]->freq == 1) {
            ans.push_back(words[i].substr(0, k));
            break;
          }
          temp = temp->chars[index];
        }
      }
    }
    return ans;
  }
};

int main() {
  int n = 3;
  string words[] = {"geeksgeeks", "geeksquiz", "geeksforgeeks"};
  Trie t = Trie();
  for (int i = 0; i < n; i++) t.insert(words[i]);

  vector<string> ans = t.unique_prefixes(words, n);
  for (auto x : ans) {
    cout << x << " ";
  }
}

File: tries/lb-06-all-shortest-unique-prefixes.py
class TrieNode:
    def __init__(self):
        self.chars = [[0,0]]*26 # Node and count of words starting with this node
        self.terminal = False


class Trie:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()

    def get_idx(self, w):
        return ord(w) - ord('a')

    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        root = self.root
        for w in word:
            index = self.get_idx(w)
            if not root.chars[index][0]:
                root.chars[index] = [TrieNode(), 0]
            # already present then increase count and move ahead
            root.chars[index][1]+=1
            root = root.chars[index][0]
        root.terminal = True
    
    def unique_prefixes(self, words):
        ans = []
        def helper(root, word):
            for i, w in enumerate(word):
                index = self.get_idx(w)
                if root.chars[index][0]:
                    # only one word from this node then it must be the prefix
                    if root.chars[index][1]==1:
                        ans.append(word[:i+1])
                        break
                    root = root.chars[index][0]
                # else not present in the tree

          
        for w in words:          
            helper(self.root, w)
            
        return ans
                
    
t = Trie()
# c = ["zebra", "dog", "duck", "dove"]
c = ["geeksgeeks", "geeksquiz", "geeksforgeeks"]
for i in c:
    t.insert(i)
print(t.unique_prefixes(c))

# Note Trie can also be modified to store the count of each word staring with char
# to reduce the constant searching time each time for each node = O(nk)

# Giving Memory Error on Large input?